### PR TITLE
Refactor loggerd rotations

### PR DESCRIPTION
**Pull request by ZwX1616**
so that all hevcs have perfectly matching frames - tici prep
first segments will still be a bit off since we want to start logging as soon as going onroad

credit: @deanlee

### CODE REVIEW

**Comment by deanlee** [comment]
@ZwX1616 , Frankly, I spent a lot of time trying to understand the code, I'm confused by the current complex.
maybe I don’t really understand what problem we are trying to solve?

**Comment by ZwX1616** [comment]
> @ZwX1616 , Frankly, I spent a lot of time trying to understand the code, I'm confused by the current complex.
> maybe I don’t really understand what problem we are trying to solve?

making sure frames are perfectly matched(synced) between cameras

File: selfdrive/loggerd/loggerd.cc [hunk 1]
```diff
@@ -86,75 +133,119 @@ volatile sig_atomic_t do_exit = 0;
...
-  std::condition_variable cv;
   char segment_path[4096];
-  uint32_t last_frame_id;
-  uint32_t rotate_last_frame_id;
   int rotate_segment;
-  int rotate_seq_id;
-
   pthread_mutex_t rotate_lock;
   int num_encoder;
+  int rotate_seq_id;
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:195]: multiple threads access.
should be declared using the std::atomic<> .

File: selfdrive/loggerd/loggerd.cc [hunk 2]
```diff
@@ -203,66 +303,46 @@ void encoder_thread(bool is_streaming, bool raw_clips, int cam_idx) {
...
-        if (encoder.rotating) {
           pthread_mutex_lock(&s.rotate_lock);
           s.should_close += 1;
           pthread_mutex_unlock(&s.rotate_lock);
 
-          while(s.should_close > 0 && s.should_close < s.num_encoder) {
-            // printf("%d waiting for others to reach close, %d/%d \n", my_idx, s.should_close, s.num_encoder);
-            s.cv.wait(lk);
-          }
+          while(s.should_close > 0 && s.should_close < s.num_encoder && !do_exit) { usleep(1000); }
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:341]: read should be protected too.

```
std::atomic<int> should_close;
...
...
s.should_close+=1;
while (s.should_close < s.num_encoder && !do_exit) { usleep(1000); }
```

File: selfdrive/loggerd/loggerd.cc [hunk 3]
```diff
@@ -86,75 +133,119 @@ volatile sig_atomic_t do_exit = 0;
...
-  uint32_t last_frame_id;
-  uint32_t rotate_last_frame_id;
   int rotate_segment;
-  int rotate_seq_id;
-
   pthread_mutex_t rotate_lock;
   int num_encoder;
+  int rotate_seq_id;
   int should_close;
   int finish_close;
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:197]: std::atomic<int> finish_close

File: selfdrive/loggerd/loggerd.cc [hunk 4]
```diff
@@ -276,13 +356,18 @@ void encoder_thread(bool is_streaming, bool raw_clips, int cam_idx) {
           s.finish_close += 1;
           pthread_mutex_unlock(&s.rotate_lock);
 
-          while(s.finish_close > 0 && s.finish_close < s.num_encoder) {
-            // printf("%d waiting for others to actually close, %d/%d \n", my_idx, s.finish_close, s.num_encoder);
-            s.cv.wait(lk);
-          }
+          while(s.finish_close > 0 && s.finish_close < s.num_encoder && !do_exit) { usleep(1000); }
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:359]: It seems that incoming camera frames may be lost while waiting for other encoders to complete. will this be a problem?

File: selfdrive/loggerd/loggerd.cc [hunk 5]
```diff
@@ -203,66 +303,46 @@ void encoder_thread(bool is_streaming, bool raw_clips, int cam_idx) {
...
 
           pthread_mutex_lock(&s.rotate_lock);
-          if (s.should_close == s.num_encoder) {
-            s.should_close = 1 - s.num_encoder;
-          } else {
-            s.should_close += 1;
-          }
+          s.should_close = s.should_close == s.num_encoder ? 1 - s.num_encoder : s.should_close + 1;
+
           encoder_close(&encoder);
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:346]: encoder_close needs time to flush out omx buffers.maybe we can consider putting the call before waiting s.should_close == s.num_encoder

File: selfdrive/loggerd/loggerd.cc (hunk 4 above)
Comment (deanlee) [reply selfdrive/loggerd/loggerd.cc:359]: sorry,forget about it.I got it wrong, it's yuv stream from pool, not rgb from tbuffer

Comment (deanlee) [reply selfdrive/loggerd/loggerd.cc:359]: modeld may also encounter the same problem. https://github.com/commaai/openpilot/issues/2146

File: selfdrive/loggerd/loggerd.cc [hunk 6]
```diff
@@ -86,75 +132,119 @@ volatile sig_atomic_t do_exit = 0;
...
+  void cancelWait() { cv.notify_one(); }
+
+  void setStreamFrameId(uint32_t frame_id) {
+    fid_lock.lock();
+    stream_frame_id = frame_id;
+    fid_lock.unlock();
+    cv.notify_one();
+  }
+
+  void setLogFrameId(uint32_t frame_id) {
```
Comment (deanlee) [inline selfdrive/loggerd/loggerd.cc:163]: is this function necessary?

**Pull request merged**